### PR TITLE
Add auto bp peering functionality

### LINKED ISSUE #716: Nodeos option to auto-peer based on BP schedule
**User Story:** As a Block Producer, I want a nodeos option to auto-peer based on the BP schedule, so that I can minimize latency with the node before me in the schedule and avoid late starts.

**Context:** A Block Producer may get a late start in building blocks for a round, due to latency in receiving blocks from the BP that falls prior to them in the BP schedule. When a BP gets a late start, they risk being forced to miss a block or broadcast an empty block.

To optimize latency and minimize late starts, BPs want a reliable way to:
- receive the last block of the previous round as soon as possible
- send their last block to the next Block Producer as soon as possible

So, in addition to the manual peer configuration, there should be a configuration option to automatically peer based on the BP schedule to minimize late starts.

**Requirements:**
- "BP P2P Config" an out-of-band configuration correlating BP accounts to P2P connection info (a list of endpoints)
- Establish a separate bp-schedule-max-peer limit of 2
- Should be an option "BP Schedule Auto Peering" (on/off)
- If BP Schedule Auto Peering is on:
  - Allow connections from nodes in BP P2P Config, up to the bp-schedule-max-peer limit, giving priority based on schedule   - proximity
  - When Block Producer Schedule changes attempt to connect to the next 2 BPs in schedule.
  - When schedule changes, if BP that is no longer in schedule proximity is in your manual peer list, do not disrupt the connection.
  - When schedule changes, if BP becomes in schedule proximity, but is already connected, leave the connection alone... and when they are no longer in proximity, do not disconnect.
  - BP Schedule Auto Peering should not conflict with manual Peer configuration settings
  - BP Schedule Auto Peering should not consider or apply to max peers, unless the BP is in your manual peer configuration

### PR DESCRIPTION
**Pull request by huangminghuang**
This PR address https://github.com/AntelopeIO/leap/issues/716 to provide the functionality to automatically create p2p connections based on  the BP schedule.

The option `--p2p-auto-bp-peer` is added to associate the BP account name and its p2p address. When a block producer detects a pending schedule change, it would look up the **next two** block producers on the new schedule configured in `--p2p-auto-bp-peer` list and establish the p2p connections to the those neighboring producers if found. When a BP scheduling is actually changed, it would tear down any existing auto created p2p connection which are no longer in its scheduling proximity. In addition, these auto created connections are not subjected to the limit specified by the `--max-clients` option. 

Resolves #716 

The output of `v1/net/connections` api endpoint would add a `is_bp_peer` to every connection object. An example would be
```
[
  {
    "peer": "",
    "connecting": false,
    "syncing": false,
    "is_bp_peer": true,
    "last_handshake": {
      "network_version": 1212,
      "chain_id": "f0d5b776137d7760c0085646114fa81907c6c5f47203e981ea0a5d7d2197c15c",
      "node_id": "2f80c7136cdfb9dbc9b0265308bad52b440814963dfecfb1fca9d96b103e9152",
      "key": "EOS1111111111111111111111111111111114T1Anm",
      "time": "1677632015959642628",
      "token": "0000000000000000000000000000000000000000000000000000000000000000",
      "sig": "SIG_K1_111111111111111111111111111111111111111111111111111111111111111116uk5ne",
      "p2p_address": "localhost:9894 - 2f80c71",
      "last_irreversible_block_num": 214,
      "last_irreversible_block_id": "000000d65790e7fb47478d996076af5e940e43f75b6a578bbfc9400cedd153f5",
      "head_num": 331,
      "head_id": "0000014be4b87dd168e9a1b75b4d72dd22fda7bbbc830893affd02e8c215c7a5",
      "os": "linux",
      "agent": "EOS Test Agent",
      "generation": 1
    }
  },
  {
    "peer": "localhost:9776",
    "connecting": false,
    "syncing": false,
    "is_bp_peer": false,
    "last_handshake": {
      "network_version": 1212,
      "chain_id": "f0d5b776137d7760c0085646114fa81907c6c5f47203e981ea0a5d7d2197c15c",
      "node_id": "aa782bac601bc45fb61e3b01bc1c9a722b0ecd722a79376db2b934038ab063ce",
      "key": "EOS1111111111111111111111111111111114T1Anm",
      "time": "1677631870325641251",
      "token": "0000000000000000000000000000000000000000000000000000000000000000",
      "sig": "SIG_K1_111111111111111111111111111111111111111111111111111111111111111116uk5ne",
      "p2p_address": "localhost:9776 - aa782ba",
      "last_irreversible_block_num": 46,
      "last_irreversible_block_id": "00000029229cabd541d97cc0bb43de0496dea44b04df689acc9e82fe858b5c9f",
      "head_num": 48,
      "head_id": "0000002b8bd6c46477b983c30a48caba056b5df17a7b804f634eccfe5cbb6c1d",
      "os": "linux",
      "agent": "EOS Test Agent",
      "generation": 1
    }
  },
  {
    "peer": "localhost:9877",
    "connecting": false,
    "syncing": false,
    "is_bp_peer": true,
    "last_handshake": {
      "network_version": 1212,
      "chain_id": "f0d5b776137d7760c0085646114fa81907c6c5f47203e981ea0a5d7d2197c15c",
      "node_id": "bc3f55b501c0cd082c6cc146d81898e9ecb019aa3b425a1e0007296079f28c2e",
      "key": "EOS1111111111111111111111111111111114T1Anm",
      "time": "1677632015962242909",
      "token": "0000000000000000000000000000000000000000000000000000000000000000",
      "sig": "SIG_K1_111111111111111111111111111111111111111111111111111111111111111116uk5ne",
      "p2p_address": "localhost:9877 - bc3f55b",
      "last_irreversible_block_num": 214,
      "last_irreversible_block_id": "000000d65790e7fb47478d996076af5e940e43f75b6a578bbfc9400cedd153f5",
      "head_num": 331,
      "head_id": "0000014be4b87dd168e9a1b75b4d72dd22fda7bbbc830893affd02e8c215c7a5",
      "os": "linux",
      "agent": "EOS Test Agent",
      "generation": 1
    }
  },
  {
    "peer": "localhost:9876",
    "connecting": false,
    "syncing": false,
    "is_bp_peer": true,
    "last_handshake": {
      "network_version": 1212,
      "chain_id": "f0d5b776137d7760c0085646114fa81907c6c5f47203e981ea0a5d7d2197c15c",
      "node_id": "46e2911bc9c1d49ae9aeb407d94ab0634fd675145de278866ae31dd7faf1d493",
      "key": "EOS1111111111111111111111111111111114T1Anm",
      "time": "1677632015960414749",
      "token": "0000000000000000000000000000000000000000000000000000000000000000",
      "sig": "SIG_K1_111111111111111111111111111111111111111111111111111111111111111116uk5ne",
      "p2p_address": "localhost:9876 - 46e2911",
      "last_irreversible_block_num": 214,
      "last_irreversible_block_id": "000000d65790e7fb47478d996076af5e940e43f75b6a578bbfc9400cedd153f5",
      "head_num": 331,
      "head_id": "0000014be4b87dd168e9a1b75b4d72dd22fda7bbbc830893affd02e8c215c7a5",
      "os": "linux",
      "agent": "EOS Test Agent",
      "generation": 1
    }
  },
  {
    "peer": "",
    "connecting": false,
    "syncing": false,
    "is_bp_peer": true,
    "last_handshake": {
      "network_version": 1212,
      "chain_id": "f0d5b776137d7760c0085646114fa81907c6c5f47203e981ea0a5d7d2197c15c",
      "node_id": "35824e4246973d5863f48870008af6a53b9db479905b3fa742b5a1978f4ae406",
      "key": "EOS1111111111111111111111111111111114T1Anm",
      "time": "1677632015966290027",
      "token": "0000000000000000000000000000000000000000000000000000000000000000",
      "sig": "SIG_K1_111111111111111111111111111111111111111111111111111111111111111116uk5ne",
      "p2p_address": "localhost:9895 - 35824e4",
      "last_irreversible_block_num": 214,
      "last_irreversible_block_id": "000000d65790e7fb47478d996076af5e940e43f75b6a578bbfc9400cedd153f5",
      "head_num": 331,
      "head_id": "0000014be4b87dd168e9a1b75b4d72dd22fda7bbbc830893affd02e8c215c7a5",
      "os": "linux",
      "agent": "EOS Test Agent",
      "generation": 1
    }
  }
]
```

### CODE REVIEW

File: plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+using chain::flat_set;
+
+template <typename Derived, typename Connection>
+class connection_manager {
+#ifdef BOOST_TEST_MODULE
+ public:
+#endif
+
+   flat_map<account_name, std::string> bp_peer_addresses;
+   flat_map<std::string, account_name> bp_peer_accounts;
```
Comment (heifner) [inline plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp:23]: `bp_peer_addresses` and `bp_peer_accounts` are only modified on plugin_startup so they are currently thread safe to view from any threads. However, if we add ability to modify the configuration without restarting nodeos these will need some kind of thread protection. A comment about thread safety would be nice.

**Comment by heifner** [comment]
@bhazzard Should the `--p2p-auto-bp-peer` be something that can be modified via an http endpoint? It complicates the implementation but I can see it being desired. Or is that something that should wait on: https://github.com/AntelopeIO/leap/issues/103

File: plugins/net_plugin/net_plugin.cpp [hunk 2]
```diff
@@ -3702,14 +3748,24 @@ namespace eosio {
...
          }
          if( options.count( "agent-name" )) {
             my->user_agent_name = options.at( "agent-name" ).as<string>();
             EOS_ASSERT( my->user_agent_name.length() <= max_handshake_str_length, chain::plugin_config_exception,
                         "agent-name too long, must be less than ${m}", ("m", max_handshake_str_length) );
          }
 
+         if ( options.count( "p2p-auto-bp-peer")) {
+            my->set_bp_peers(options.at( "p2p-auto-bp-peer" ).as<vector<string>>());
+            my->for_each_bp_peer_address([this](auto addr) {
```
Comment (heifner) [inline plugins/net_plugin/net_plugin.cpp:3762]: This can be `const auto& addr`

File: plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+   // This should only be called after the first handshake message is received to check if an incoming connection
+   // has exceeded the pre-configured max_client_count limit.
+   bool exceeding_connection_limit(Connection* new_connection) const {
+      return auto_bp_peering_enabled() && self()->max_client_count != 0 &&
+             established_client_connection(new_connection) && num_established_clients() > self()->max_client_count;
+   }
+
+   struct neighbor_finder_type {
+      const flat_map<account_name, std::string>&    bp_peer_addresses;
+      const flat_set<account_name>&                 my_accounts;
```
Comment (heifner) [inline plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp:108]: Not needed as not used outside of the constructor.

File: plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+   }
+
+   // This should only be called after the first handshake message is received to check if an incoming connection
+   // has exceeded the pre-configured max_client_count limit.
+   bool exceeding_connection_limit(Connection* new_connection) const {
+      return auto_bp_peering_enabled() && self()->max_client_count != 0 &&
+             established_client_connection(new_connection) && num_established_clients() > self()->max_client_count;
+   }
+
+   struct neighbor_finder_type {
```
Comment (heifner) [inline plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp:106]: This struct could be made private except of BOOST_TEST.

**Review by heifner — CHANGES_REQUESTED** [review]
This review limited to thread-safety. I did not review any of the logic or funcationality.

File: plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+      }
+      return result.str();
+   }
+
+   class neighbor_finder_type {
+
+      const flat_map<account_name, std::string>&    bp_peer_addresses;
+      const flat_set<account_name>&                 my_accounts;
+      const std::vector<chain::producer_authority>& schedule;
+      chain::flat_set<std::size_t>                  indeces;
```
Comment (greg7mdp) [inline plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp:49]:
```suggestion
      chain::flat_set<std::size_t>                  indices;
```

File: plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+    public:
+      neighbor_finder_type(const flat_map<account_name, std::string>&    bp_peer_addresses,
+                           const flat_set<account_name>&                 my_accounts,
+                           const std::vector<chain::producer_authority>& schedule)
+          : bp_peer_addresses(bp_peer_addresses), my_accounts(my_accounts), schedule(schedule) {
+         for (auto account : my_accounts) {
+            auto itr = std::find_if(schedule.begin(), schedule.end(),
+                                    [account](auto& e) { return e.producer_name == account; });
```
Comment (greg7mdp) [inline plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp:57]: This has O(n<sup>2</sup>) complexity. Maybe not a big issue if there aren't many peers. 
Still would be preferable to build a `flat_mat<account_name, account_name>` mapping `producer_name` -> `account`, to replace the O(n) find_if in the `std::vector`.

Comment (greg7mdp) [reply plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp:57]: Sounds good then, thanks for letting me know the expectation.

File: libraries/chain/include/eosio/chain/producer_schedule.hpp [hunk 7]
```diff
@@ -335,4 +335,3 @@ FC_REFLECT_DERIVED( eosio::chain::producer_schedule_change_extension, (eosio::ch
 FC_REFLECT( eosio::chain::shared_block_signing_authority_v0, (threshold)(keys))
 FC_REFLECT( eosio::chain::shared_producer_authority, (producer_name)(authority) )
 FC_REFLECT( eosio::chain::shared_producer_authority_schedule, (version)(producers) )
-
```
Comment (dimas1185) [inline libraries/chain/include/eosio/chain/producer_schedule.hpp:338]: unnecessary line removal?

File: plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+   flat_set<account_name>              active_neighbors;
+   uint32_t                            pending_schedule_version = 0;
+   uint32_t                            active_schedule_version  = 0;
+
+   Derived*       self() { return static_cast<Derived*>(this); }
+   const Derived* self() const { return static_cast<const Derived*>(this); }
+
+   template <template<typename...> typename Container, typename... Rest >
+   static std::string to_string(const Container<account_name, Rest...>& peers) {
+      std::stringstream result;
```
Comment (dimas1185) [inline plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp:35]: not sure if better but here is one liner example that should do same:
```c++
return boost::algorithm::join(peers | boost::adaptors::transformed([](auto& p){ return p.to_string(); }), ",");
```

File: plugins/net_plugin/net_plugin.cpp [hunk 9]
```diff
@@ -2747,7 +2774,7 @@ namespace eosio {
             boost::system::error_code ec;
             acceptor->cancel( ec );
             acceptor->close( ec );
-         }
```
Comment (dimas1185) [inline plugins/net_plugin/net_plugin.cpp:2750]: please avoid whitespace only changes.
same comment for lines 2570-2579

File: plugins/net_plugin/net_plugin.cpp [hunk 10]
```diff
@@ -2864,7 +2899,7 @@ namespace eosio {
                      // we need to avoid the case where they would both tell a different connection to go away.
                      // Using the sum of the initial handshake times of the two connections, we will
                      // arbitrarily (but consistently between the two peers) keep one of them.
-                  
+
```
Comment (dimas1185) [inline plugins/net_plugin/net_plugin.cpp:2902]: whitespace only change

File: plugins/net_plugin/net_plugin.cpp [hunk 11]
```diff
@@ -2882,7 +2917,7 @@ namespace eosio {
                               ("lhs", my_impl->node_id)("rhs", msg.node_id) );
                      // only the connection from lower node_id to higher node_id will be considered as a duplicate,
                      // so there is no chance for both connections to be closed
-                     continue; 
+                     continue;
```
Comment (dimas1185) [inline plugins/net_plugin/net_plugin.cpp:2920]: whitespace only change

File: plugins/net_plugin/net_plugin.cpp [hunk 12]
```diff
@@ -3872,9 +3937,9 @@ namespace eosio {
    void net_plugin::plugin_shutdown() {
       try {
          fc_ilog( logger, "shutdown.." );
-         
+
          my->plugin_shutdown();
-         
+
```
Comment (dimas1185) [inline plugins/net_plugin/net_plugin.cpp:3942]: whitespace only change (same as line 3875)

File: testnet.template [hunk 13]
```diff
@@ -15,7 +15,7 @@ fi
 bioscontractpath=$BIOS_CONTRACT_PATH
 if [ -z "$bioscontractpath" ]; then
     # this is defaulted to the version of bios that only requires the preactivate_feature
-    # newer versions may require 
+    # newer versions may require
```
Comment (dimas1185) [inline testnet.template:18]: whitespace only change

File: plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+      }
+      return result.str();
+   }
+
+   class neighbor_finder_type {
+
+      const flat_map<account_name, std::string>&    bp_peer_addresses;
+      const flat_set<account_name>&                 my_accounts;
+      const std::vector<chain::producer_authority>& schedule;
+      chain::flat_set<std::size_t>                  indices;
```
Comment (dimas1185) [inline plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp:49]: name `indices` doesn't sound self descriptive. I suggest to rename it to something like `my_schedule_indices` or something like that.

File: plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+         result << sep << p.to_string();
+         sep = ", ";
+      }
+      return result.str();
+   }
+
+   class neighbor_finder_type {
+
+      const flat_map<account_name, std::string>&    bp_peer_addresses;
+      const flat_set<account_name>&                 my_accounts;
```
Comment (dimas1185) [inline plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp:47]: `my_accounts` doesn't sound self-descriptive. I would name it something like `my_bp_accounts` or `self_bp_accounts`

File: plugins/net_plugin/net_plugin.cpp (hunk 9 above)
Comment (huangminghuang) [reply plugins/net_plugin/net_plugin.cpp:2750]: I am sorry. My code editor does removing trailing spaces automatically, I never try to configure it otherwise because no one else has ever had problem with it. However, I don't think it makes sense to add them back because it won't preserve the history unless I do a force push which is frowned upon by the team.

File: plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+   void set_bp_peers(const std::vector<std::string>& peers) {
+      try {
+         for (auto& entry : peers) {
+            auto comma_pos = entry.find(',');
+            EOS_ASSERT(comma_pos != std::string::npos, chain::plugin_config_exception,
+                       "auto-bp-peer must consists an account name and server address separated by a comma token");
+            auto         addr = entry.substr(comma_pos + 1);
+            account_name account(entry.substr(0, comma_pos));
+
+            config.bp_peer_addresses[account] = addr;
```
Comment (heifner) [inline plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp:111]: `std::move(addr)`

File: plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+            config.bp_peer_accounts[addr]     = account;
+         }
+      } catch (eosio::chain::name_type_exception&) {
```
Comment (heifner) [inline plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp:114]: This try-catch could be inside the for loop so it could report which account string(s) are invalid.

File: plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+                       ("old", pending_schedule_version)("new", schedule.version));
+
+               auto finder                       = neighbor_finder(schedule.producers);
+               auto pending_downstream_neighbors = finder.downstream_neighbors();
+
+               fc_dlog(self()->get_logger(), "pending_downstream_neighbors: ${pending_downstream_neighbors}",
+                       ("pending_downstream_neighbors", to_string(pending_downstream_neighbors)));
+               for (auto neighbor : pending_downstream_neighbors) { self()->connect(config.bp_peer_addresses[neighbor]); }
+
+               pending_neighbors = pending_downstream_neighbors;
```
Comment (heifner) [inline plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp:183]: `std::move(pending_downstream_neighbors)`

File: plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+               /// establish connection to the BPs within our pending scheduling proximity
+
+               fc_dlog(self()->get_logger(), "pending producer schedule switches from version ${old} to ${new}",
+                       ("old", pending_schedule_version)("new", schedule.version));
+
+               auto finder                       = neighbor_finder(schedule.producers);
+               auto pending_downstream_neighbors = finder.downstream_neighbors();
+
+               fc_dlog(self()->get_logger(), "pending_downstream_neighbors: ${pending_downstream_neighbors}",
+                       ("pending_downstream_neighbors", to_string(pending_downstream_neighbors)));
```
Comment (heifner) [inline plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp:180]: Is there a reason to only log downstream neighbors and not upstream neighbors?

File: plugins/net_plugin/include/eosio/net_plugin/net_plugin.hpp [hunk 20]
```diff
@@ -10,6 +10,7 @@ namespace eosio {
       string            peer;
       bool              connecting = false;
       bool              syncing    = false;
+      bool              is_bp_peer = false;
```
Comment (heifner) [inline plugins/net_plugin/include/eosio/net_plugin/net_plugin.hpp:13]: Note this in the PR description and provide example output.

File: plugins/net_plugin/net_plugin.cpp [hunk 21]
```diff
@@ -2842,6 +2870,14 @@ namespace eosio {
          }
 
          log_p2p_address = msg.p2p_address;
+
+         my_impl->mark_bp_connection(this);
+         if (my_impl->exceeding_connection_limit(this)) {
+            fc_dlog(logger, "max_client_count ${m} exceeded", ("m", my_impl->max_client_count));
+            my_impl->disconnect(peer_address());
+            return;
+         }
```
Comment (heifner) [inline plugins/net_plugin/net_plugin.cpp:2879]: Why was this added? Why is the check in `start_listen_loop` not sufficient?

File: plugins/net_plugin/net_plugin.cpp [hunk 22]
```diff
@@ -3627,6 +3673,11 @@ namespace eosio {
            "    p2p.blk.eos.io:9876:blk\n")
          ( "p2p-max-nodes-per-host", bpo::value<int>()->default_value(def_max_nodes_per_host), "Maximum number of client nodes from any single IP address")
          ( "p2p-accept-transactions", bpo::value<bool>()->default_value(true), "Allow transactions received over p2p network to be evaluated and relayed if valid.")
+         ( "p2p-auto-bp-peer", bpo::value< vector<string> >()->composing(),
+           "The account and public p2p endpoint of a block producer node to automatically connect to when the it is in producer schedule proximity\n."
+           "   Syntax: account,host:port\n"
+           "   Example,\n"
+           "     eosproducer1,p2p.eos.io:9876\n")
```
Comment (heifner) [inline plugins/net_plugin/net_plugin.cpp:3680]: Add an example that is block only which in practice they should all be specified as block only.
`eosproducer2,p2p.eos2.io:9876:blk`
It looks like your code supports using `blk` and `trx`, but please verify.

File: plugins/net_plugin/tests/auto_bp_peering_unittest.cpp [hunk 23]
```diff
@@ -0,0 +1,288 @@
...
+      }
+   }
+
+   std::function<void(std::string)> connect;
+   std::function<void(std::string)> disconnect;
+
+   void setup_test_peers() {
+      set_bp_peers({ "proda,127.0.0.1:8001"s, "prodb,127.0.0.1:8002"s, "prodc,127.0.0.1:8003"s, "prodd,127.0.0.1:8004"s,
+                     "prode,127.0.0.1:8005"s, "prodf,127.0.0.1:8006"s, "prodg,127.0.0.1:8007"s, "prodh,127.0.0.1:8008"s,
+                     "prodi,127.0.0.1:8009"s, "prodj,127.0.0.1:8010"s,
```
Comment (heifner) [inline plugins/net_plugin/tests/auto_bp_peering_unittest.cpp:39]: Add some `:blk` and `:trx` peers.

**Review by heifner — CHANGES_REQUESTED** [review]

File: plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp (hunk 19 above)
Comment (huangminghuang) [reply plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp:180]: Because those the only peers to make connections to.

File: plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp (hunk 17 above)
Comment (huangminghuang) [reply plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp:114]: I believe https://github.com/AntelopeIO/leap/blob/48a4d871db92fa00767490994aab526f4c0db857/libraries/chain/name.cpp#L13 already report the invalid string

File: plugins/net_plugin/net_plugin.cpp (hunk 21 above)
Comment (heifner) [reply plugins/net_plugin/net_plugin.cpp:2879]: Can you add a comment with this info.

**Review by heifner — APPROVED** [review]

File: plugins/net_plugin/net_plugin.cpp (hunk 9 above)
Comment (dimas1185) [reply plugins/net_plugin/net_plugin.cpp:2750]: got it. try to avoid in future, please. it breaks changes history and adds verbosity into PR

**Review by dimas1185 — APPROVED** [review]

File: plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp [hunk 24]
```diff
@@ -0,0 +1,230 @@
...
+   }
+
+   // Only called from connection strand and the connection constructor
+   void mark_bp_connection(Connection* conn) const {
+      /// mark an connection as a bp connection if it connects to an address in the bp peer list, so that the connection
+      /// won't be subject to the limit of max_client_count.
+      auto space_pos = conn->log_p2p_address.find(' ');
+      // log_p2p_address always has a trailing hex like `localhost:9877 - bc3f55b`
+      std::string addr = conn->log_p2p_address.substr(0, space_pos);
+      if (auto itr = config.bp_peer_accounts.find(addr); itr != config.bp_peer_accounts.end()) {
```
Comment (greg7mdp) [inline plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp:134]:
```suggestion
      if (config.bp_peer_accounts.count(addr)) {
```
